File: GeeksforGeeks_POTD/July_2026/July_8.py
"""
For p and q seperately make boolean matrix of all cells  - Can be reached by that tower or not according to conditions(By BFS)
Then, find common in both --> ANS

TC - O(nm), SC - O(nm)
"""

class Solution:
    def countCoordinates(self, mat):
        m = len(mat)
        n = len(mat[0])
    
        p_queue = []
        q_queue = []
        
        for i in range(m):
            p_queue.append((i,0))
            q_queue.append((i,n-1))
        
        for i in range(n):
            p_queue.append((0,i))
            q_queue.append((m-1,i))
        
        
        def tower_reach_by_station(queue, visit):
            
            while queue:
                r,c = queue.pop(0)
                if (r,c) in visit:
                    continue
                
                visit.add((r,c))
                
                for dr,dc in [(0,1), (0,-1), (1,0), (-1,0)]:
                    nr,nc = dr+r, dc+c
                    if (nr, nc) not in visit and 0 <= nr < m and 0<= nc < n and mat[nr][nc] >= mat[r][c]:
                        queue.append((nr,nc))
            
            return visit
        
        
        tower_reach_by_p = tower_reach_by_station(p_queue, set())
        tower_reach_by_q = tower_reach_by_station(q_queue, set())

        return len(tower_reach_by_p.intersection(tower_reach_by_q))

File: GeeksforGeeks_POTD/July_2026/July_8.java
package GeeksforGeeks_POTD.July_2026;

/*
For p and q seperately make boolean matrix of all cells  - Can be reached by that tower or not according to conditions(By BFS)
Then, find common in both --> ANS

TC - O(nm), SC - O(nm)
*/

import java.util.ArrayDeque;
import java.util.Queue;

public class July_8 {
    public int countCoordinates(int[][] mat) {
        int n=mat.length;
        int m=mat[0].length;
        int ans=0;
        boolean[][] p=new boolean[n][m];
        Queue<int[]> a=new ArrayDeque<>();
        for(int i=0;i<n;i++) a.offer(new int[]{i,0});
        for(int i=1;i<m;i++) a.offer(new int[]{0,i});
        while(!a.isEmpty()){
            int[] ac=a.poll();
            if(!p[ac[0]][ac[1]]){
                p[ac[0]][ac[1]]=true;
                if(ac[1]+1<m &&  mat[ac[0]][ac[1]]<=mat[ac[0]][ac[1]+1]) a.offer(new int[]{ac[0],ac[1]+1});
                if(ac[1]-1>=0 &&  mat[ac[0]][ac[1]]<=mat[ac[0]][ac[1]-1]) a.offer(new int[]{ac[0],ac[1]-1});
                if(ac[0]+1<n &&  mat[ac[0]][ac[1]]<=mat[ac[0]+1][ac[1]]) a.offer(new int[]{ac[0]+1,ac[1]});
                if(ac[0]-1>=0 &&  mat[ac[0]][ac[1]]<=mat[ac[0]-1][ac[1]]) a.offer(new int[]{ac[0]-1,ac[1]});
            }
        }
        boolean[][] q=new boolean[n][m];
        for(int i=0;i<n;i++) a.offer(new int[]{i,m-1});
        for(int i=0;i<m-1;i++) a.offer(new int[]{n-1,i});
        while(!a.isEmpty()){
            int[] ac=a.poll();
            if(!q[ac[0]][ac[1]]){
                q[ac[0]][ac[1]]=true;
                if(p[ac[0]][ac[1]]) ans++;
                if(ac[1]+1<m &&  mat[ac[0]][ac[1]]<=mat[ac[0]][ac[1]+1]) a.offer(new int[]{ac[0],ac[1]+1});
                if(ac[1]-1>=0 &&  mat[ac[0]][ac[1]]<=mat[ac[0]][ac[1]-1]) a.offer(new int[]{ac[0],ac[1]-1});
                if(ac[0]+1<n &&  mat[ac[0]][ac[1]]<=mat[ac[0]+1][ac[1]]) a.offer(new int[]{ac[0]+1,ac[1]});
                if(ac[0]-1>=0 &&  mat[ac[0]][ac[1]]<=mat[ac[0]-1][ac[1]]) a.offer(new int[]{ac[0]-1,ac[1]});
            }
        }
        return ans;
    }
}
